pipeline_tools/gcs_utils.py

```diff
@@ -0,0 +1,108 @@
+"""This module contains utility functions and classes to interact with Google Cloud Storage Service.
+"""
+import logging
+from google.cloud import storage
+import google.auth
+from io import BytesIO
+
+
+def get_filename_from_gs_link(link):
+    """Get the filename corresponding to a google_storage link.
+
+    :param str link: A string of google cloud storage link.
+    :return str: A string of filename.
+    """
+    return link.split('/')[-1]
+
+
+def parse_bucket_blob_from_gs_link(path):
+    """Utility to split a google storage path into bucket + blob name.
+
+    :param str path: A string of google cloud storage path (must have gs:// prefix)
+    :return str: A string of bucket name.
+    :return str: A string of blob name
+    """
+    if not path.startswith('gs://'):
+        raise ValueError('%s path is not a valid link')
+    parts = path.split('/', 3)
+    (prefix, _, bucket), blob = parts[:3], parts[3]
+
+    return bucket, blob
+
+
+def download_to_buffer(blob):
+    """Return a bytes file-like object readable by requests and REST APIs
+
+    :param google.cloud.storage.Blob blob: google storage blob
+    :return _io.BufferedIOBase: readable file object
+    """
+    string_buffer = BytesIO()
+    blob.download_to_file(string_buffer)
+    string_buffer.seek(0)
+    return string_buffer
+
+
+def download_gcs_blob(gcs_client, bucket_name, source_blob_name):
+    """Use google.cloud.storage API to download a blob from the bucket.
+
+    :param GoogleCloudStorageClient gcs_client: A GoogleCloudStorageClient object with a
+            google.cloud.storage.client.Client instance as a lazy-initialized property.
+    :param str bucket_name: A string of bucket name.
+    :param str source_blob_name: A string of source blob name that to be downloaded.
+    :return BufferedIOBase: File-like object returned by download_to_buffer.
+    """
+    logging.getLogger()
+
+    # Make sure the lazy property storage_client of gcs_client instance initialized at least once
+    if not gcs_client.storage_client:
+        gcs_client.storage_client
+    authenticated_gcs_client = gcs_client.storage_client
+
+    bucket = authenticated_gcs_client.bucket(bucket_name)
+    blob = bucket.blob(source_blob_name)
+    logging.debug('bucket: {0}, blob: {1}'.format(bucket_name, source_blob_name))
+    return download_to_buffer(blob)
+
+
+class LazyProperty(object):
```

Comment:
@rexwangcc Is this replaceable with python's `functools.lru_cache` with the max_size parameter set to 1?

Reply:
I did a quick research about the [functools implementation](https://github.com/python/cpython/blob/60a376cf0023d1070329d0e861a5596637ff3275/Lib/functools.py#L399) of LRU cache. I think it is possible we could replace the current decorator with the lru_cache decorator. To do this, we need to do one of the following two things: 
1. We could still use the GoogleCloudStorageClient class, but we need to implement the `__hash__` method, to enable lru_cache to work for the client, we might need to change the way that passing credentials to the client, those lists are not hashable. In short, in this way, we make the whole GoogleCloudStorageClient class hashable, and decorate the download_gcs_blob with lru_cache, since the max size is 1 and it is thread-safe, it will always hit the cache. 
2. A neater way requires more refactoring work: refactor the code and remove the GoogleCloudStorageClient class. Instead, we just need to use storage_client function and decorate the download_gcs_blob. Since we are using storage_client to store the information of the credentials, this will also ACT-AS the lazy property, and will only throw errors during the runtime.

For the second way, I tested it a little bit, codes are shown as below(alert me if there are any credentials I put in the code):
```python
from google.cloud import storage
from google.oauth2 import service_account
from io import BytesIO
from functools import lru_cache

def download_to_buffer(blob):
    string_buffer = BytesIO()
    blob.download_to_file(string_buffer)
    string_buffer.seek(0)
    return string_buffer

@lru_cache(maxsize=1)
def download_gcs_blob(bucket_name, source_blob_name):
    authenticated_gcs_client = storage_client()
    bucket = authenticated_gcs_client.bucket(bucket_name)
    blob = bucket.blob(source_blob_name)
    return download_to_buffer(blob)


def storage_client(key_location='/etc/secondary-analysis/bucket-reader-key.jon', scopes=['https://www.googleapis.com/auth/devstorage.read_only']):
    credentials = service_account.Credentials.from_service_account_file(
        key_location, scopes=scopes)
    client = storage.Client(credentials=credentials, project=credentials.project_id)
    return client

for i in range(40):
    wdlIO = download_gcs_blob("broad-dsde-mint-dev-teststorage", "prepare_and_analyze_mock_smartseq2.wdl")

download_gcs_blob.cache_info()
"""Out[2]: CacheInfo(hits=39, misses=1, maxsize=1, currsize=1)"""
```

The result looks reasonable, if we gonna use this build-in decorator, we could benefit from the official support for the decorator as well as the fast velocity of C-implementation of the LRU cache algorithm, and this could be more scalable than our current lazy property. But whether to change or not depends on where we still want to download wdls from, Github or buckets. Overall, thank you very much for this idea! @ambrosejcarr

Reply:
Interesting idea. It would be nice to simplify this code. Let's make a separate ticket for this.